src/Workspaces/Remote/Core/SolutionAsset.cs

```diff
@@ -5,31 +5,51 @@
 #nullable enable
 
 using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Serialization;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     /// <summary>
-    /// Asset that is part of solution
+    /// Represents a part of solution snapshot along with its checksum.
     /// </summary>
-    internal sealed class SolutionAsset : RemotableData
+    internal sealed class SolutionAsset
     {
-        private readonly object _value;
-        private readonly ISerializerService _serializer;
+        public static readonly SolutionAsset Null = new SolutionAsset(value: null, Checksum.Null, WellKnownSynchronizationKind.Null);
 
-        public SolutionAsset(Checksum checksum, object value, ISerializerService serializer)
-            : base(checksum, value.GetWellKnownSynchronizationKind())
+        /// <summary>
+        /// Indicates what kind of object it.
+        /// 
+        /// Used in tranportation framework and deserialization service
+        /// to hand shake how to send over data and deserialize serialized data.
+        /// </summary>
+        public readonly WellKnownSynchronizationKind Kind;
+
+        /// <summary>
+        /// Checksum of <see cref="Value"/>.
+        /// </summary>
+        public readonly Checksum Checksum;
+
+        public readonly object? Value;
+
+        public SolutionAsset(object? value, Checksum checksum, WellKnownSynchronizationKind kind)
+        {
+            Checksum = checksum;
+            Kind = kind;
+            Value = value;
+        }
+
+        public SolutionAsset(Checksum checksum, object value)
+            : this(value, checksum, value.GetWellKnownSynchronizationKind())
         {
-            _value = value;
-            _serializer = serializer;
         }
 
-        public override Task WriteObjectToAsync(ObjectWriter writer, CancellationToken cancellationToken)
+        public void WriteObjectTo(ObjectWriter writer, ISerializerService serializer, CancellationToken cancellationToken)
         {
-            _serializer.Serialize(_value, writer, cancellationToken);
-            return Task.CompletedTask;
+            if (Value is not null)
+            {
+                serializer.Serialize(Value, writer, cancellationToken);
```

Comment:
can you doc.  this feels odd.  i.e. i would still expect to write out a null value so the other side sees that.

Reply:
Null is already written in the stream as Checksum.Null and WellKnownSynchronizationKind.Null. I'll inline this method because it does not make sense on it's own. The serialization format can definitely use some improvements.